src/gui/sdlmain.cpp

```diff
@@ -1388,77 +1381,66 @@ void GFX_SwitchFullScreen(void) {
  	GFX_ResetScreen();
 }
 
-static void SwitchFullScreen(bool pressed) {
-	if (!pressed)
-		return;
-
-	if (sdl.desktop.lazy_fullscreen) {
-//		sdl.desktop.lazy_fullscreen_req=true;
-		LOG_MSG("GFX LF: fullscreen switching not supported");
-	} else {
+static void SwitchFullScreen(bool pressed)
+{
+	if (pressed)
 		GFX_SwitchFullScreen();
-	}
-}
-
-void GFX_SwitchLazyFullscreen(bool lazy) {
-	sdl.desktop.lazy_fullscreen=lazy;
-	sdl.desktop.lazy_fullscreen_req=false;
 }
 
-void GFX_SwitchFullscreenNoReset(void) {
-	sdl.desktop.fullscreen=!sdl.desktop.fullscreen;
-}
-
-bool GFX_LazyFullscreenRequested(void) {
-	if (sdl.desktop.lazy_fullscreen) return sdl.desktop.lazy_fullscreen_req;
-	return false;
-}
-
-bool GFX_StartUpdate(Bit8u * & pixels,Bitu & pitch) {
+// This function returns write'able buffer for user to draw upon. Successful
+// return depends on properly initialized SDL_Block structure (which generally
+// can be achieved via GFX_SetSize call), and specifically - properly initialized
+// output-specific bits (sdl.surface, sdl.texture, sdl.opengl.framebuf, or
+// sdl.openg.pixel_buffer_object fields).
+// 
+// If everything is prepared correctly, this function returns true, assigns 
+// 'pixels' output parameter to to a buffer (with format specified via earlier
+// GFX_SetSize call), and assigns 'pitch' to a number of bytes used for a single
+// pixels row in 'pixels' buffer.
+//
+bool GFX_StartUpdate(uint8_t * &pixels, int &pitch)
+{
 	if (!sdl.update_display_contents)
 		return false;
 	if (!sdl.active || sdl.updating)
 		return false;
+
 	switch (sdl.desktop.type) {
-	case SCREEN_SURFACE:
-		pixels = (Bit8u *)sdl.surface->pixels;
-		pixels += sdl.clip.y * sdl.surface->pitch;
-		pixels += sdl.clip.x * sdl.surface->format->BytesPerPixel;
-		pitch = sdl.surface->pitch;
-		sdl.updating = true;
-		return true;
 	case SCREEN_TEXTURE:
-	{
-		void * texPixels;
-		int texPitch;
-		if (SDL_LockTexture(sdl.texture.texture, NULL, &texPixels, &texPitch) < 0)
-			return false;
-		pixels = (Bit8u *)texPixels;
-		pitch = texPitch;
-		sdl.updating=true;
+		pixels = static_cast<uint8_t *>(sdl.texture.input_surface->pixels);
+		pitch = sdl.texture.input_surface->pitch;
```

Comment:
`assert(sdl.texture.input_surface)` before 1410 to prevent null dereference? (not sure if it needed though).

Reply:
Yes! :) Using these functions in a new context without assertions and without documentation was very inconvenient. I added all asserts suggested below as well.